src/Grpc.AspNetCore.Server/Internal/ServerCallDeadlineManager.cs

```diff
@@ -0,0 +1,252 @@
+        }
+
+        private void RequestAborted()
+        {
+            // Call is complete if the request has aborted
+            lock (this)
+            {
+                IsCallComplete = true;
+                _deadlineCts!.Cancel();
+            }
+        }
+
+        private long GetTimerDueTime(TimeSpan timeout, long maxTimerDueTime)
+        {
+            // Timer has a maximum allowed due time.
+            // The called method will rechedule the timer if the deadline time has not passed.
+            var dueTimeMilliseconds = timeout.Ticks / TimeSpan.TicksPerMillisecond;
+            dueTimeMilliseconds = Math.Min(dueTimeMilliseconds, maxTimerDueTime);
+            // Timer can't have a negative due time
+            dueTimeMilliseconds = Math.Max(dueTimeMilliseconds, 0);
+
+            return dueTimeMilliseconds;
+        }
+
+        public void SetCallEnded()
+        {
+            lock (this)
+            {
+                IsCallComplete = true;
+            }
+        }
+
+        public bool TrySetCallComplete()
+        {
+            lock (this)
+            {
+                if (!IsDeadlineExceededStarted)
+                {
+                    IsCallComplete = true;
+                    return true;
+                }
+
+                return false;
+            }
+        }
+
+        private async Task DeadlineExceededAsync()
+        {
+            if (!TryStartExceededDeadline())
+            {
+                return;
+            }
+
+            try
+            {
+                await _serverCallContext.DeadlineExceededAsync();
+                lock (this)
+                {
+                    IsCallComplete = true;
+                    _deadlineCts?.Cancel();
+                }
+            }
+            finally
+            {
+                _deadlineExceededCompleteTcs!.TrySetResult(null);
+            }
+        }
+
+        private bool TryStartExceededDeadline()
+        {
+            lock (this)
+            {
+                // Deadline callback could be raised by the CTS after call has been completed (either successfully, with error, or aborted)
+                // but before deadline exceeded registration has been disposed
+                if (!IsCallComplete)
+                {
+                    _deadlineExceededCompleteTcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    return true;
+                }
+
+                return false;
+            }
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            // Deadline registration needs to be disposed with DisposeAsync, and the task completed
+            // before the lock can be disposed.
+            // Awaiting deadline registration and deadline task ensures it has finished running, so there is
+            // no way for deadline logic to attempt to wait on a disposed lock.
+            var disposeTask = _longDeadlineTimer.DisposeAsync();
```

Comment:
```c#
private static void DeadlineExceededCallback(object? state) => _ = ((ServerCallDeadlineManager)state!).DeadlineExceededAsync();
```

Because the DeadlineExceededCallback discards the Task returned by DeadlineExceededAsync(), awaiting Timer.DisposeAsync() does not ensure that DeadlineExceededAsync() has finished running. _deadlineCts could get disposed before a pending call to Cancel() causing an unobserved ODE.

Since TimerCallback isn't Task-returning, I think [using a Timer-backed awaitable](https://github.com/dotnet/aspnetcore/blob/8c8cb4dce235b099d1e419b6a039217eb56a77e6/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs#L1792-L1816) is the way to go.

Reply:
The manager gets around that problem by using a TCS - _deadlineExceededCompleteTcs. If it is set then DisposeAsync will await it before disposing the CTS.

However, now that I think about it, I could just avoid using a Timer altogether and use a Task.Delay. A deadline only happens once.

```cs
private async Task StartDeadlineAsync(TimeSpan time)
{
    await Task.Delay(time);
    await DeadlineExceededAsync();
}

public ServerCallDeadlineManager(...)
{
    _deadlineTask = StartDeadlineAsync(...);
}
```

Reply:
Hmm, but a CTS would be needed to stop the `Task.Delay`, and there would be an `OCE`.

I think the current approach would be more efficient.

Reply:
Another option is to have the TimerCallback complete yet another TCS. You could then await that TCS's Task at the start of DeadlineExceededAsync(), invoke DeadlineExceededAsync() in the ServerCallDeadlineManager ctor, and await the Task returned by DeadlineExceededAsync() in DisposeAsync().

Reply:
Is _deadlineExceededCompleteTcs guaranteed to be set (if it will be set) before DisposeAsync is called?